core/libp2p/crypto/key_generator/key_generator_impl.cpp

```diff
@@ -31,14 +32,48 @@ namespace libp2p::crypto {
   }
 
   namespace detail {
-    KeyGenerator::Buffer bio2buffer(BIO *bio) {
-      auto length = BIO_pending(bio);
-      std::vector<uint8_t> bytes(length + 1, 0);
-      BIO_read(bio, bytes.data(), length);
+    // encode cryptographic key in `ASN.1 DER` format
+    template <class KeyStructure, class Function>
+    outcome::result<std::vector<uint8_t>> encodeKeyDer(KeyStructure *ks,
+                                                       Function *function) {
```

Comment:
Why not std::function to specify the function signature and avoid pointers?

Reply:
the `function` argument can receive function pointer of different signatures, so I cannot transform it into `std::function<signature>` with fixed signature.